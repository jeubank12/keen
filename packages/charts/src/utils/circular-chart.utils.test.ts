import {
  generateCircularChart,
  createStackedSlice,
} from './circular-chart.utils';

import { pieChart } from '../charts/pie/pie-chart.fixtures';

describe('@keen/charts - circular chart utils', () => {
  describe('createStackedSlice()', () => {
    const slices = [
      { color: 'red', value: 100, selector: [0] },
      { color: 'blue', value: 100, selector: [1] },
      { color: 'yellow', value: 5, selector: [2] },
      { color: 'violet', value: 3, selector: [3] },
      { color: 'green', value: 2, selector: [4] },
    ];

    it('should create "stack" from 3 slices', () => {
      const result = createStackedSlice({ slices, treshold: 5, total: 210 });

      expect(result).toMatchSnapshot();
    });

    it('should not create "stack" and return original slices', () => {
      const result = createStackedSlice({ slices, treshold: 1, total: 210 });

      expect(result).toMatchSnapshot();
    });
  });

  describe('generateCircularChart()', () => {
    it('should create arcs based on provided data', () => {
      const { arcs } = generateCircularChart(pieChart);
      const { data } = pieChart;

      expect(arcs.length).toEqual(data.length);
    });

    it('should calculate properties for each arc', () => {
      const { arcs } = generateCircularChart(pieChart);

      expect(arcs).toMatchSnapshot();
    });

    it('should calculate position for labels placed "outer" of slice', () => {
      const { arcs } = generateCircularChart({
        ...pieChart,
        labelsPosition: 'outside',
      });

      expect(arcs).toMatchSnapshot();
    });
  });
});
