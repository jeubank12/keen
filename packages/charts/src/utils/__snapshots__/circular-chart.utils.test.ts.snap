// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@keen/charts - circular chart utils createStackedSlice() should create "stack" from 3 slices 1`] = `
Array [
  Object {
    "color": "red",
    "selector": Array [
      0,
    ],
    "value": 100,
  },
  Object {
    "color": "blue",
    "selector": Array [
      1,
    ],
    "value": 100,
  },
  Object {
    "color": "#CDCFD3",
    "selector": Array [],
    "stack": Array [
      Object {
        "color": "yellow",
        "selector": Array [
          2,
        ],
      },
      Object {
        "color": "violet",
        "selector": Array [
          3,
        ],
      },
      Object {
        "color": "green",
        "selector": Array [
          4,
        ],
      },
    ],
    "stacked": true,
    "value": 10,
  },
]
`;

exports[`@keen/charts - circular chart utils createStackedSlice() should not create "stack" and return original slices 1`] = `
Array [
  Object {
    "color": "red",
    "selector": Array [
      0,
    ],
    "value": 100,
  },
  Object {
    "color": "blue",
    "selector": Array [
      1,
    ],
    "value": 100,
  },
  Object {
    "color": "yellow",
    "selector": Array [
      2,
    ],
    "value": 5,
  },
  Object {
    "color": "violet",
    "selector": Array [
      3,
    ],
    "value": 3,
  },
  Object {
    "color": "green",
    "selector": Array [
      4,
    ],
    "value": 2,
  },
]
`;

exports[`@keen/charts - circular chart utils generateCircularChart() should calculate position for labels placed "outer" of slice 1`] = `
Array [
  Object {
    "activePosition": Array [
      -7.397584510798206,
      -6.728725243729192,
    ],
    "color": "#85B4C3",
    "endAngle": 5.866829654294192,
    "index": "4",
    "label": "13.3%",
    "labelPosition": Array [
      -51.783091575587434,
      -47.101076706104344,
    ],
    "selector": Array [
      0,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 5.034118348523403,
  },
  Object {
    "activePosition": Array [
      -3.341041714973898,
      9.425361545256726,
    ],
    "color": "#CB5623",
    "endAngle": 4.087855501056598,
    "index": "2",
    "label": "19.3%",
    "labelPosition": Array [
      -23.387292004817287,
      65.97753081679708,
    ],
    "selector": Array [
      1,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 2.876639056299088,
  },
  Object {
    "activePosition": Array [
      6.7287252437291905,
      -7.3975845107982074,
    ],
    "color": "#E29B1E",
    "endAngle": 1.4761700420482162,
    "index": "0",
    "label": "23.5%",
    "labelPosition": Array [
      47.10107670610433,
      -51.78309157558745,
    ],
    "selector": Array [
      2,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 0,
  },
  Object {
    "activePosition": Array [
      -9.885605855918886,
      1.5082429716137402,
    ],
    "color": "#487650",
    "endAngle": 5.034118348523403,
    "index": "3",
    "label": "15.1%",
    "labelPosition": Array [
      -69.19924099143219,
      10.557700801296182,
    ],
    "selector": Array [
      3,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 4.087855501056598,
  },
  Object {
    "activePosition": Array [
      8.2215601169481,
      5.692622352080662,
    ],
    "color": "#27566D",
    "endAngle": 2.876639056299088,
    "index": "1",
    "label": "22.3%",
    "labelPosition": Array [
      57.55092081863669,
      39.84835646456463,
    ],
    "selector": Array [
      4,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 1.4761700420482162,
  },
  Object {
    "activePosition": Array [
      -2.981979110466657,
      -9.54504062771555,
    ],
    "color": "#674AA9",
    "endAngle": 6.093932737686226,
    "index": "5",
    "label": "3.6%",
    "labelPosition": Array [
      -20.8738537732666,
      -66.81528439400886,
    ],
    "selector": Array [
      5,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 5.866829654294192,
  },
  Object {
    "activePosition": Array [
      -0.9448513188803126,
      -9.95526272808559,
    ],
    "color": "#F4A083",
    "endAngle": 6.283185307179586,
    "index": "6",
    "label": "3.0%",
    "labelPosition": Array [
      -6.613959232162188,
      -69.68683909659912,
    ],
    "selector": Array [
      6,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 6.093932737686226,
  },
]
`;

exports[`@keen/charts - circular chart utils generateCircularChart() should calculate properties for each arc 1`] = `
Array [
  Object {
    "activePosition": Array [
      -7.397584510798206,
      -6.728725243729192,
    ],
    "color": "#85B4C3",
    "endAngle": 5.866829654294192,
    "index": "4",
    "label": "13.3%",
    "labelPosition": Array [
      -33.28913029859192,
      -30.279263596781366,
    ],
    "selector": Array [
      0,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 5.034118348523403,
  },
  Object {
    "activePosition": Array [
      -3.341041714973898,
      9.425361545256726,
    ],
    "color": "#CB5623",
    "endAngle": 4.087855501056598,
    "index": "2",
    "label": "19.3%",
    "labelPosition": Array [
      -15.03468771738254,
      42.41412695365527,
    ],
    "selector": Array [
      1,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 2.876639056299088,
  },
  Object {
    "activePosition": Array [
      6.7287252437291905,
      -7.3975845107982074,
    ],
    "color": "#E29B1E",
    "endAngle": 1.4761700420482162,
    "index": "0",
    "label": "23.5%",
    "labelPosition": Array [
      30.279263596781355,
      -33.289130298591935,
    ],
    "selector": Array [
      2,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 0,
  },
  Object {
    "activePosition": Array [
      -9.885605855918886,
      1.5082429716137402,
    ],
    "color": "#487650",
    "endAngle": 5.034118348523403,
    "index": "3",
    "label": "15.1%",
    "labelPosition": Array [
      -44.485226351634985,
      6.787093372261831,
    ],
    "selector": Array [
      3,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 4.087855501056598,
  },
  Object {
    "activePosition": Array [
      8.2215601169481,
      5.692622352080662,
    ],
    "color": "#27566D",
    "endAngle": 2.876639056299088,
    "index": "1",
    "label": "22.3%",
    "labelPosition": Array [
      36.997020526266446,
      25.616800584362977,
    ],
    "selector": Array [
      4,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 1.4761700420482162,
  },
  Object {
    "activePosition": Array [
      -2.981979110466657,
      -9.54504062771555,
    ],
    "color": "#674AA9",
    "endAngle": 6.093932737686226,
    "index": "5",
    "label": "3.6%",
    "labelPosition": Array [
      -13.418905997099957,
      -42.95268282471998,
    ],
    "selector": Array [
      5,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 5.866829654294192,
  },
  Object {
    "activePosition": Array [
      -0.9448513188803126,
      -9.95526272808559,
    ],
    "color": "#F4A083",
    "endAngle": 6.283185307179586,
    "index": "6",
    "label": "3.0%",
    "labelPosition": Array [
      -4.251830934961407,
      -44.79868227638516,
    ],
    "selector": Array [
      6,
    ],
    "stack": undefined,
    "stacked": undefined,
    "startAngle": 6.093932737686226,
  },
]
`;
